packages/common/src/font.ts

```diff
@@ -136,58 +130,156 @@ export const getFontKitFont = async (textSchema: TextSchema, font: Font) => {
   }
 
   const fontKitFont = fontkit.create(fontData instanceof Buffer ? fontData : Buffer.from(fontData as ArrayBuffer));
-  fontKitFontCache[fontName] = fontKitFont
+  fontKitFontCache[fontName] = fontKitFont;
 
   return fontKitFont;
-}
+};
+
+export type FontWidthCalcValues = {
+  font: FontKitFont;
+  fontSize: number;
+  characterSpacing: number;
+  boxWidthInPt: number;
+};
+
+const isTextExceedingBoxWidth = (text: string, calcValues: FontWidthCalcValues) => {
+  const { font, fontSize, characterSpacing, boxWidthInPt } = calcValues;
+  const textWidth = widthOfTextAtSize(text, font, fontSize, characterSpacing);
+  return textWidth >= boxWidthInPt;
+};
+
+/**
+ * Incrementally checks the current line for its real length
+ * and returns the position where it exceeds the box width.
+ * Returns `null` to indicate if textLine is shorter than the available box.
+ */
+const getOverPosition = (textLine: string, calcValues: FontWidthCalcValues) => {
+  for (let i = 0; i <= textLine.length; i++) {
+    if (isTextExceedingBoxWidth(textLine.slice(0, i + 1), calcValues)) {
+      return i;
+    }
+  }
+
+  return null;
+};
+
+/**
+ * Gets the position of the split. Splits the exceeding line at
+ * the last whitespace prior to it exceeding the bounding box width.
+ */
+const getSplitPosition = (textLine: string, calcValues: FontWidthCalcValues) => {
+  const overPos = getOverPosition(textLine, calcValues);
+  if (overPos === null) return textLine.length; // input line is shorter than the available space
+
+  let overPosTmp = overPos;
+  while (textLine[overPosTmp] !== ' ' && overPosTmp >= 0) {
+    overPosTmp--;
+  }
 
-export const calculateDynamicFontSize = async ({ textSchema, font, input, }: { textSchema: TextSchema; font: Font; input: string; }) => {
-  const { fontSize: _fontSize, dynamicFontSize: dynamicFontSizeSetting, characterSpacing, width, height, lineHeight = DEFAULT_LINE_HEIGHT } = textSchema;
-  const fontSize = _fontSize || DEFAULT_FONT_SIZE;
+  // For very long lines with no whitespace use the original overPos
+  return overPosTmp > 0 ? overPosTmp : overPos - 1;
+};
+
+/**
+ * Recursively splits the line at getSplitPosition.
+ * If there is some leftover, split the rest again in the same manner.
+ */
+export const getSplittedLines = (textLine: string, calcValues: FontWidthCalcValues): string[] => {
```

Comment:
Nice.

Reply:
@peteward 
Wait, moved it to "common," but isn't it being used in the UI?

This is fine for now, but if it's not going to be used commonly, wouldn't it have been better to move it to a different file?

If we split functions, it might become difficult to delete them in the future, and there's a possibility that they could become unused functions. Additionally, considering the Custom Schema that we will implement in the future, I think it would be better to gather all the processes related to this TextSchema in one place.

Reply:
Originally this function was only used in the generator. Now it is used by both the generator and the UI, so I moved it to common/font.
Let me know if you think there is a more appropriate place.

Reply:
Ah, sorry, it's being used in 'calculateDynamicFontSize', I missed that.